Sources/AsyncHTTPClient/FileDownloadDelegate.swift

```diff
@@ -0,0 +1,88 @@
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+import NIOHTTP1
+
+/// Handles a streaming download to a given file path, allowing headers and progress to be reported.
+public final class FileDownloadDelegate: HTTPClientResponseDelegate {
+    /// The response type for this delegate: the total count of bytes as reported by the response
+    /// "Content-Length" header (if available) and the count of bytes downloaded.
+    public typealias Response = (totalBytes: Int?, receivedBytes: Int)
+
+    private var totalBytes: Int?
+    private var receivedBytes = 0
+
+    private let handle: NIOFileHandle
+    private let io: NonBlockingFileIO
+    private let reportHeaders: ((HTTPHeaders) -> Void)?
+    private let reportProgress: ((_ totalBytes: Int?, _ receivedBytes: Int) -> Void)?
+
+    private var writeFuture: EventLoopFuture<Void>?
+
+    /// Initializes a new file download delegate.
+    /// - parameters:
+    ///     - path: Path to a file you'd like to write the download to.
+    ///     - pool: A thread pool to use for asynchronous file I/O.
+    ///     - reportHeaders: A closure called when the response headers are available.
+    ///     - reportProgress: A closure called when a body chunk has been downloaded, with
+    ///       the total byte count and download byte count passed to it as arguments.
+    public init(
+        path: String,
+        pool: NIOThreadPool = NIOThreadPool(numberOfThreads: 1),
+        reportHeaders: ((HTTPHeaders) -> Void)? = nil,
+        reportProgress: ((_ totalBytes: Int?, _ receivedBytes: Int) -> Void)? = nil
+    ) throws {
+        self.handle = try NIOFileHandle(path: path, mode: .write, flags: .allowFileCreation())
+        pool.start()
+        self.io = NonBlockingFileIO(threadPool: pool)
+
+        self.reportHeaders = reportHeaders
+        self.reportProgress = reportProgress
+    }
+
+    public func didReceiveHead(
+        task: HTTPClient.Task<Response>,
+        _ head: HTTPResponseHead
+    ) -> EventLoopFuture<Void> {
+        self.reportHeaders?(head.headers)
+
+        if let totalBytesString = head.headers.first(name: "Content-Length"),
+            let totalBytes = Int(totalBytesString) {
+            self.totalBytes = totalBytes
+        }
+
+        return task.eventLoop.makeSucceededFuture(())
+    }
+
+    public func didReceiveBodyPart(
+        task: HTTPClient.Task<Response>,
+        _ buffer: ByteBuffer
+    ) -> EventLoopFuture<Void> {
+        self.receivedBytes += buffer.readableBytes
+        self.reportProgress?(self.totalBytes, self.receivedBytes)
+
+        let writeFuture = self.io.write(fileHandle: self.handle, buffer: buffer, eventLoop: task.eventLoop)
+        self.writeFuture = writeFuture
+        return writeFuture
+    }
+
+    public func didFinishRequest(task: HTTPClient.Task<Response>) throws -> Response {
+        self.writeFuture?.whenComplete { [weak self] _ in
```

Comment:
Capturing `self` weakly here is an anti-pattern: if the user drops the `FileDownloadDelegate` such that it is deinited, we will never close the handle and so we will leak it. We must capture self strongly here.

More generally, there is little risk when capturing self strongly in one-shot closures. The reason to use a weak capture is when we are passing self to something that we ourselves hold. Future callbacks, like dispatch work items, are transient: when they have been invoked they no longer exist and are freed. It's almost always safe to capture self strongly in those contexts.

Reply:
I am aware of this reasoning, the problem is that the future itself references this closure, which captures `self`. There is no need for `[weak self]` if this is a one-shot closure, but since it's owned by `self` it creates a circular reference: `self` -> `writeFuture` -> closure -> `self`.

Reply:
Strong reference cycles are not a problem so long as they are deterministically broken. In this case, this future callback _must_ complete: it is a program error for it to fail to do so. Thus, this closure will be executed, and once it's executed the future will drop its reference to it and the cycle will break.

Reply:
Cool, makes sense. I didn't see this behavior documented anywhere, either for `EventLoopFuture` specifically, or in general in related NIO docs. I had enough experience with both 3rd-party and system libraries that don't follow this behavior and also don't document it in any way, so I'm used to staying on the safe side.

Reply:
apple/swift-nio#1575.

Reply:
That's great, thanks!